src/main/scala/is/hail/utils/HashMethods.scala

```diff
@@ -32,3 +32,50 @@ class TwoIndepHash32(outBits: Int, a: Long, b: Long) extends (Int => Int) {
 
   override def apply(key: Int): Int = ((a * key + b) >>> shift).toInt
 }
+
+object SimpleTabulationHash32 {
+  def apply(rand: RandomDataGenerator): SimpleTabulationHash32 = {
+    new SimpleTabulationHash32(Array.fill[Int](256 * 4)(rand.getRandomGenerator.nextInt()))
+  }
+}
+
+// see e.g. Thorup, "Fast and Powerful Hashing using Tabulation", for explanation of both simple and twisted tabulation
+// hashing
+class SimpleTabulationHash32(table: Array[Int]) extends (Int => Int) {
+  require(table.length == 256 * 4)
+
+  override def apply(key: Int): Int = {
+    var out: Int = 0
+    var keyBuffer: Int = key
+    var offset = 0
+    while (offset < 1024) {
+      out ^= table(offset + (keyBuffer & 0xff))
+      offset += 256
```

Comment:
Just a remark, the optimization that changes:

```
var i = 0
while (i < 4) {
    ... i * 256 ...
   i += 1
}
```

to:

```
  while (offset < 1024) {
    ... offset ...
    offset += 256
}
```

is called strength reduction and it is a bog-standard compiler optimization, so you should feel free to write the former if you think it is more readable.